### PR TITLE
HW 19. Introduction to monitoring. Monitoring systems (monitoring-1)

### PR DESCRIPTION
**Pull request by gitpushforcemain**
# Completed Homework №19

- [x] :large_blue_diamond: Main tasks
- [x] :large_orange_diamond: Extra tasks for self-study

## The following tasks have been completed

- :large_blue_diamond: Created GCE instance `docker-host` using `docker-machine` command for the Reddit and Prometheus containers
- :large_blue_diamond: Launched Prometheus container built from `prom/prometheus:v2.1.0` to learn about the UI and basic Prometheus metrics
- :large_blue_diamond: Created new image `$(USER_NAME)/prometheus` built using `prom/prometheus:v2.1.0` with custom config provided in `prometheus.yml`
- :large_blue_diamond: Built images for the Reddit app services using `docker_build.sh` scripts provided in `ui`, `post-py`, and `comment` folders
- :large_blue_diamond: Added `prometheus` service to the Compose file and launched microservices using `docker-compose up -d` command
- :large_blue_diamond: Using Prometheus web interface inspected services' health status, stopped some services to check Prometheus status codes and relaunched the services again to make sure that all services are healthy
- :large_blue_diamond: Added `node-exporter` service to the Compose file, added `node` job to the `prometheus.yml`, and rebuilt `$(USER_NAME)/prometheus` image
- :large_blue_diamond: Pushed images used in this homework to the Docker Hub
- :large_orange_diamond: Added Dockerfile to `docker/mongodb_exporter` folder to build Docker image and added `mongodb` job to `prometheus.yml` for MongoDB monitoring service
- :large_orange_diamond: Added Dockerfile to `docker/blackbox_exporter` folder to build Docker image and added `blackbox` job with `http_2xx` module to `prometheus.yml` to monitor HTTP for `comment`, `ui`, and `post` microservices
- :large_orange_diamond: Created Makefile that can build all images need for this homework and can push them to Docker Hub

## How to get project running

1. Create GCE instance using `docker-machine` command
2. Run `eval $(docker-machine env docker-host)`
3. Build images with `make`
4. Run `docker-compose up -d` from `docker` folder

## How to check that project actually works

- Ensure that the build passed TravisCI tests
  - [#andrei_bogomja](https://devops-team-otus.slack.com/messages/C9M6L1XB3) Slack channel
  - actual GitHub pull request page
- Get the IP address of Prometheus service with `docker-machine ip docker-host` and make sure that all targets have `UP` status

## PR checklist

- [x] Added label with homework number
- [x] Added label with homework subject

### CODE REVIEW

File: monitoring/mongodb_exporter/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,14 @@
+FROM golang:latest as builder
+
+ENV SRC $GOPATH/src/github.com/percona/mongodb_exporter
+
+WORKDIR ${SRC}
+
+RUN git clone "https://github.com/percona/mongodb_exporter" "$SRC" \
```
Comment (Nklya) [inline monitoring/mongodb_exporter/Dockerfile:7]: Версии лучше бы фиксировать.

File: monitoring/blackbox_exporter/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+ENV SRC $GOPATH/src/github.com/prometheus/blackbox_exporter
+
+WORKDIR ${SRC}
+
+RUN git clone "https://github.com/prometheus/blackbox_exporter" "$SRC" \
+    && go get -d \
+    && go build -o blackbox_exporter
+
+FROM golang:latest
```
Comment (Nklya) [inline monitoring/blackbox_exporter/Dockerfile:11]: Для второго стейджа можно уже не использовать образ golang, т.к. билдить уже не нужно.

File: monitoring/blackbox_exporter/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,16 @@
+FROM golang:latest as builder
```
Comment (Nklya) [inline monitoring/blackbox_exporter/Dockerfile:1]: Можно было использовать готовый образ `prom/blackbox_exporter`, чтобы не билдить свой.

File: monitoring/prometheus/prometheus.yml [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+  - job_name: 'blackbox'
+    metrics_path: /probe
+    params:
+      module: [http_2xx]
+    static_configs:
+      - targets:
+        - http://ui:9292
+        - http://comment:9292
+        - http://post:9090
```
Comment (Nklya) [inline monitoring/prometheus/prometheus.yml]: У comment и post не реализован ответ на `/`, лучше прописывать `/healthcheck`

**Review by Nklya — APPROVED** [review]
Добрый вечер.

Андрей, условия ДЗ выполнены, все реализовано хорошо 👍

**Comment by Nklya** [comment]
Этот PR заапрувлен, надо бы его смержить и подмержить потом мастер в monitoring-2

**Comment by gitpushforcemain** [comment]
Смерджил в monitoring-2

**Pull request merged**